Fix memory leaks in vector and map converters #795

cmastalli opened the pull request:
This PR fixes the memory leak produced in the Python interface reported in #795.
I have test myself that there is no more memory leaks in vector and map converters.
To validate it, I used the following code:
```python
import numpy as np
import crocoddyl
import pinocchio
import example_robot_data
import time


def test1():
  model = crocoddyl.ActionModelUnicycle()
  model.r = [10,1]
  u0 = np.array([0., 0.])
  x0 = np.array([-1., -1., 1.])
  data = model.createData()
#  problem = crocoddyl.ShootingProblem(x0, [model], model, [data], data)
  duration = 0.
  while duration < 10.:
    t0 = time.time()
    problem = crocoddyl.ShootingProblem(x0, [model] * 10, model, [data] * 10, data)
    ddp = crocoddyl.SolverDDP(problem)
    ddp.solve([x0] * 11, [u0] * 10, 1)
    ddp.xs
    tf = time.time()
    duration += tf - t0

def test3():
  talos_arm = example_robot_data.loadTalosArm()
  robot_model = talos_arm.model

  state = crocoddyl.StateMultibody(robot_model)
  runningCostModel = crocoddyl.CostModelSum(state)
  terminalCostModel = crocoddyl.CostModelSum(state)
  Mref = crocoddyl.FramePlacement(robot_model.getFrameId("gripper_left_joint"),
                                  pinocchio.SE3(np.eye(3), np.array([.0, .0, .4])))
  goalTrackingCost = crocoddyl.CostModelFramePlacement(state, Mref)
  xRegCost = crocoddyl.CostModelState(state)
  uRegCost = crocoddyl.CostModelControl(state)

  runningCostModel.addCost("gripperPose", goalTrackingCost, 1)
  runningCostModel.addCost("xReg", xRegCost, 1e-4)
  runningCostModel.addCost("uReg", uRegCost, 1e-4)
  terminalCostModel.addCost("gripperPose", goalTrackingCost, 1)

  actuationModel = crocoddyl.ActuationModelFull(state)
  dt = 1e-3
  runningModel = crocoddyl.IntegratedActionModelEuler(
    crocoddyl.DifferentialActionModelFreeFwdDynamics(state, actuationModel, runningCostModel), dt)
  runningModel.differential.armature = np.array([0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.])
  terminalModel = crocoddyl.IntegratedActionModelEuler(
    crocoddyl.DifferentialActionModelFreeFwdDynamics(state, actuationModel, terminalCostModel), 0.)
  terminalModel.differential.armature = np.array([0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.])

  T = 250
  q0 = np.array([0.173046, 1., -0.52366, 0., 0., 0.1, -0.005])
  x0 = np.concatenate([q0, pinocchio.utils.zero(robot_model.nv)])

  duration = 0.
  while duration < 10.:
    t0 = time.time()
    problem = crocoddyl.ShootingProblem(x0, [runningModel] * T, terminalModel)
    problem.runningModels[0].differential.costs.costs
    tf = time.time()
    duration += tf - t0

if __name__ == '__main__':
  test1() # test2
```

and I run it using `mprof` as follows:
```
mprof run --include-children python test_memory_leak.py
mprof plot --output memory-profile.png
```

@gfadini I would be happy if you can check this yourself as well.

gfadini commented:
Hi @cmastalli, thanks I'll try the tests as soon as possible in the afternoon

gfadini commented:
Is this also adding some modification to the API? I have this error
```
  File "/home/gfadini/Desktop/pendulum_benchmark/utils.py", line 327, in plotOCSolution
    nx = xs[0].shape[0]
TypeError: No Python class registered for C++ class Eigen::Matrix<double, -1, 1, 0, -1, 1>
```

cmastalli commented:
> Why you did not use the code of Pinocchio directly? This one has a lot of redundancies, which are not useful at all.

Fixed!



> Is this also adding some modification to the API? I have this error
> 
> ```
>   File "/home/gfadini/Desktop/pendulum_benchmark/utils.py", line 327, in plotOCSolution
>     nx = xs[0].shape[0]
> TypeError: No Python class registered for C++ class Eigen::Matrix<double, -1, 1, 0, -1, 1>
> ```

Sorry, there is no expected any changes in the API. I have fixed it, please pull the latest changes @gfadini.

gfadini commented:
Ok, I'm on it

gfadini commented:
I confirm that the memory issue is solved for all the tests that I had set up. Also in the current routine of the optimization I have the memory usage has reached and is not exceeding a fixed treshold

gfadini commented:
The 19 - pybinds-solvers seems to be failing for me

cmastalli commented:
> The 19 - pybinds-solvers seems to be failing for me

All the unittests should pass now

cmastalli commented:
There is still a final problem that I need to solve. Most probably I will finish tomorrow.
